saga/sagaState.go

```diff
@@ -0,0 +1,292 @@
+				}
+			} else {
+				if !state.taskCompleted[taskId] {
+					return nil, errors.New(fmt.Sprintf("InvalidSagaState: End Saga Message cannot be applied to a Saga where Task %s has not completed", taskId))
+				}
+			}
+		}
+
+		state.sagaCompleted = true
+
+	case AbortSaga:
+		state.sagaAborted = true
+
+	case StartTask:
+		err := validateTaskId(msg.taskId)
+		if err != nil {
+			return nil, err
+		}
+
+		state.taskStarted[msg.taskId] = true
+
+	case EndTask:
+		err := validateTaskId(msg.taskId)
+		if err != nil {
+			return nil, err
+		}
+
+		// All EndTask Messages must have a preceding StartTask Message
+		if !state.taskStarted[msg.taskId] {
+			return nil, errors.New(fmt.Sprintf("Invalid Saga State: Cannot have a EndTask %s Message Before a StartTask %s Message", msg.taskId, msg.taskId))
+		}
+
+		state.taskCompleted[msg.taskId] = true
+
+	case StartCompTask:
+		err := validateTaskId(msg.taskId)
+		if err != nil {
+			return nil, err
+		}
+
+		//In order to apply compensating transactions a saga must first be aborted
+		if !state.sagaAborted {
+			return nil, errors.New(fmt.Sprintf("Invalid SagaState: Cannot have a StartCompTask %s Message when Saga has not been Aborted", msg.taskId))
+		}
+
+		// All StartCompTask Messages must have a preceding StartTask Message
+		if !state.taskStarted[msg.taskId] {
+			return nil, errors.New(fmt.Sprintf("Invalid Saga State: Cannot have a StartCompTask %s Message Before a StartTask %s Message", msg.taskId, msg.taskId))
+		}
+
+		state.compTaskStarted[msg.taskId] = true
+
+	case EndCompTask:
+		err := validateTaskId(msg.taskId)
+		if err != nil {
+			return nil, err
+		}
+
+		//in order to apply compensating transactions a saga must first be aborted
+		if !state.sagaAborted {
+			return nil, errors.New(fmt.Sprintf("Invalid SagaState: Cannot have a EndCompTask %s Message when Saga has not been Aborted", msg.taskId))
+		}
+
+		// All EndCompTask Messages must have a preceding StartTask Message
+		if !state.taskStarted[msg.taskId] {
+			return nil, errors.New(fmt.Sprintf("Invalid Saga State: Cannot have a StartCompTask %s Message Before a StartTask %s Message", msg.taskId, msg.taskId))
+		}
+
+		// All EndCompTask Messages must have a preceding StartCompTask Message
+		if !state.compTaskStarted[msg.taskId] {
+			return nil, errors.New(fmt.Sprintf("Invalid Saga State: Cannot have a EndCompTask %s Message Before a StartCompTaks %s Message", msg.taskId, msg.taskId))
+		}
+
+		state.compTaskCompleted[msg.taskId] = true
+	}
+
+	return state, nil
+}
+
+/*
+ * Creates a deep copy of mutable saga state.  Does not Deepcopy
+ * Job field since this is never mutated after creation
+ */
+func copySagaState(s *SagaState) *SagaState {
+
+	newS := &SagaState{
+		sagaId:        s.sagaId,
+		sagaAborted:   s.sagaAborted,
+		sagaCompleted: s.sagaCompleted,
+		version:       s.version + 1,
```

Comment:
Incrementing the version is not a copy. I'd make this just a copy, and then update the version in the apply function above.